tests/DCPS/UnitTests/DynamicData/DynamicData.cpp

```diff
@@ -0,0 +1,905 @@
+  ret = data.get_complex_value(nested_dd, 11);
+  EXPECT_EQ(ret, DDS::RETCODE_OK);
+  ret = nested_dd.get_float128_value(float_128, random_id);
+  EXPECT_EQ(ret, DDS::RETCODE_OK);
+  check_float128(expected.float_128, float_128);
+
+  ACE_CDR::Char char_8;
+  ret = data.get_char8_value(char_8, 12);
+  EXPECT_EQ(ret, DDS::RETCODE_OK);
+  EXPECT_EQ(expected.char_8, char_8);
+
+  ret = data.get_complex_value(nested_dd, 12);
+  EXPECT_EQ(ret, DDS::RETCODE_OK);
+  ret = nested_dd.get_char8_value(char_8, random_id);
+  EXPECT_EQ(ret, DDS::RETCODE_OK);
+  EXPECT_EQ(expected.char_8, char_8);
+
+  ACE_CDR::WChar char_16;
+  ret = data.get_char16_value(char_16, 13);
+  EXPECT_EQ(ret, DDS::RETCODE_OK);
+  EXPECT_EQ(expected.char_16, char_16);
+
+  ret = data.get_complex_value(nested_dd, 13);
+  EXPECT_EQ(ret, DDS::RETCODE_OK);
+  ret = nested_dd.get_char16_value(char_16, random_id);
+  EXPECT_EQ(ret, DDS::RETCODE_OK);
+  EXPECT_EQ(expected.char_16, char_16);
+
+  ACE_CDR::Octet byte;
+  ret = data.get_byte_value(byte, 14);
+  EXPECT_EQ(ret, DDS::RETCODE_OK);
+  EXPECT_EQ(expected.byte, byte);
+
+  ret = data.get_complex_value(nested_dd, 14);
+  EXPECT_EQ(ret, DDS::RETCODE_OK);
+  ret = nested_dd.get_byte_value(byte, random_id);
+  EXPECT_EQ(ret, DDS::RETCODE_OK);
+  EXPECT_EQ(expected.byte, byte);
+
+  ACE_CDR::Boolean bool_;
+  ret = data.get_boolean_value(bool_, 15);
+  EXPECT_EQ(ret, DDS::RETCODE_OK);
+  EXPECT_EQ(expected._cxx_bool, bool_);
+
+  ret = data.get_complex_value(nested_dd, 15);
+  EXPECT_EQ(ret, DDS::RETCODE_OK);
+  ret = nested_dd.get_boolean_value(bool_, random_id);
+  EXPECT_EQ(ret, DDS::RETCODE_OK);
+  EXPECT_EQ(expected._cxx_bool, bool_);
+
+  ACE_CDR::Long l;
+  ret = data.get_complex_value(nested_dd, 16);
+  EXPECT_EQ(ret, DDS::RETCODE_OK);
+  ret = nested_dd.get_int32_value(l, 0);
+  EXPECT_EQ(ret, DDS::RETCODE_OK);
+  EXPECT_EQ(expected.nested_struct.l, l);
+
+  XTypes::DynamicData nested_dd2;
+  ret = nested_dd.get_complex_value(nested_dd2, 0);
+  EXPECT_EQ(DDS::RETCODE_OK, ret);
+  ret = nested_dd2.get_int32_value(l, random_id);
+  EXPECT_EQ(DDS::RETCODE_OK, ret);
+  EXPECT_EQ(expected.nested_struct.l, l);
+
+  ACE_CDR::Char* str = 0;
+  ret = data.get_string_value(str, 17);
+  EXPECT_EQ(ret, DDS::RETCODE_OK);
+  EXPECT_STREQ(expected.str.in(), str);
+
+  ret = data.get_complex_value(nested_dd, 17);
+  EXPECT_EQ(ret, DDS::RETCODE_OK);
+  ret = nested_dd.get_string_value(str, random_id);
+  EXPECT_EQ(ret, DDS::RETCODE_OK);
+  EXPECT_STREQ(expected.str.in(), str);
+
+  ACE_CDR::WChar* wstr = 0;
+  ret = data.get_wstring_value(wstr, 18);
+  EXPECT_EQ(ret, DDS::RETCODE_OK);
+  EXPECT_STREQ(expected.wstr.in(), wstr);
+  CORBA::wstring_free(wstr);
+
+  wstr = 0;
+  ret = data.get_complex_value(nested_dd, 18);
+  EXPECT_EQ(ret, DDS::RETCODE_OK);
+  ret = nested_dd.get_wstring_value(wstr, random_id);
+  EXPECT_EQ(ret, DDS::RETCODE_OK);
+  EXPECT_STREQ(expected.wstr.in(), wstr);
+  CORBA::wstring_free(wstr);
+}
```

Comment:
Is there a test where members are read out of order and/or skipping members?

Reply:
That doesn't really matter because each interface `get_*_value` or `get_*_values` has a clean, new `Serializer` object, and starts from the beginning of the data stream and skips all members in front of the requested member.

Reply:
I'm not saying it's not going to work now, I'm saying it might break one day. (I might break it for all I know) It wouldn't have to be much. Something like repeat one of the previous reads at the end and add a comment that it's a test of an out of order read.

Reply:
Okay, I can add that. I was ignoring the tests that I'm pretty sure will work given the current implementation. But you're right that a different implementation might be broken by an out of order test.